users/src/test/java/io/crate/operation/user/PrivilegeValidationTest.java

```diff
@@ -0,0 +1,334 @@
+import io.crate.testing.SQLExecutor;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Collections;
+
+import static org.hamcrest.Matchers.is;
+
+public class PrivilegeValidationTest extends CrateDummyClusterServiceUnitTest {
+
+    private static final User crateUser = new User("crateUser",
+        ImmutableSet.of(User.Role.SUPERUSER),
+        Collections.EMPTY_SET);
+
+    private static final User dmlUser = new User("dmlUser",
+        Collections.EMPTY_SET,
+        ImmutableSet.of(new Privilege(Privilege.State.GRANT, Privilege.Type.DML, Privilege.Clazz.CLUSTER, null, "dmlUser")));
+
+    private static final User ddlUser = new User("ddlUser",
+        Collections.EMPTY_SET,
+        ImmutableSet.of(new Privilege(Privilege.State.GRANT, Privilege.Type.DDL, Privilege.Clazz.CLUSTER, null, "crate")));
+
+    private static final User dqlUser = new User("dqlUser",
+        Collections.EMPTY_SET,
+        ImmutableSet.of(new Privilege(Privilege.State.GRANT, Privilege.Type.DQL, Privilege.Clazz.CLUSTER, null, "crate")));
+
+    private static final User allUser = new User("allUser",
+        Collections.EMPTY_SET,
+        ImmutableSet.of(new Privilege(Privilege.State.GRANT, Privilege.Type.DQL, Privilege.Clazz.CLUSTER, null, "crate"),
+            new Privilege(Privilege.State.GRANT, Privilege.Type.DML, Privilege.Clazz.CLUSTER, null, "crate"),
+            new Privilege(Privilege.State.GRANT, Privilege.Type.DDL, Privilege.Clazz.CLUSTER, null, "crate")));
+
+    private static final User noPriviligeUser = new User("noPriviligeUser",
+        Collections.EMPTY_SET,
+        Collections.EMPTY_SET);
+
+    private SQLExecutor e;
+    private UserManager userManager;
+
+    @Before
+    public void setUpSQLExecutor() throws Exception {
+        userManager = new UserManagerService(null, null, null, clusterService);
+        TableIdent myBlobsIdent = new TableIdent(BlobSchemaInfo.NAME, "blobs");
+        e = SQLExecutor.builder(clusterService, () -> userManager)
+            .addBlobTable(TableDefinitions.createBlobTable(myBlobsIdent, clusterService))
+            .enableDefaultTables()
+            .build();
+    }
+
+    @Test
+    public void testDmlUserSelectStatementThrowsException() {
+        expectedException.expect(PermissionDeniedException.class);
+        expectedException.expectMessage(is("Missing 'DQL' Privilege for user 'dmlUser'"));
+        e.analyzer.boundAnalyze(SqlParser.createStatement("select * from sys.cluster"),
+            new SessionContext(0, Option.NONE, "doc", dmlUser), null);
+    }
+
+    @Test
+    public void testDmlUserMultiSelectStatementThrowsException() {
+        expectedException.expect(PermissionDeniedException.class);
+        expectedException.expectMessage(is("Missing 'DQL' Privilege for user 'dmlUser'"));
+        e.analyzer.boundAnalyze(SqlParser.createStatement("select * from (select id from sys.cluster) t1 where t1.id != null"),
```

Comment:
sure this one results in a MultiSourceSelect? think the subselect will be completely rewritten